File: workspaces/ws_004_qrbarcodescanner/ios/QRBarcodeScanner/Models/BarcodeSymbology.swift
//
//  BarcodeSymbology.swift
//  QRBarcodeScanner
//
//  Maps AVFoundation metadata object types to the symbologies declared
//  in the app spec, and provides the full set of types the scanner
//  listens for.
//

import AVFoundation

enum BarcodeSymbology {

    /// Every metadata object type the scanner should detect.
    /// Covers QR plus all 1D/2D barcodes listed in the app spec:
    /// EAN-13, EAN-8, UPC-A (derived from EAN-13), UPC-E, Code 128,
    /// Code 39, ITF, PDF417, Aztec, Data Matrix.
    static let supportedTypes: [AVMetadataObject.ObjectType] = [
        .qr,
        .ean13,
        .ean8,
        .upce,
        .code128,
        .code39,
        .code39Mod43,
        .code93,
        .interleaved2of5,   // ITF
        .itf14,
        .pdf417,
        .aztec,
        .dataMatrix
    ]

    /// Human-readable name for a detected metadata object type.
    ///
    /// EAN-13 payloads beginning with `0` are actually UPC-A codes;
    /// callers may pass the decoded value to refine the label.
    static func displayName(for type: AVMetadataObject.ObjectType,
                            value: String? = nil) -> String {
        switch type {
        case .qr:
            return "QR Code"
        case .ean13:
            if let v = value, v.count == 13, v.hasPrefix("0") {
                return "UPC-A"
            }
            return "EAN-13"
        case .ean8:
            return "EAN-8"
        case .upce:
            return "UPC-E"
        case .code128:
            return "Code 128"
        case .code39, .code39Mod43:
            return "Code 39"
        case .code93:
            return "Code 93"
        case .interleaved2of5:
            return "ITF"
        case .itf14:
            return "ITF-14"
        case .pdf417:
            return "PDF417"
        case .aztec:
            return "Aztec"
        case .dataMatrix:
            return "Data Matrix"
        default:
            return "Barcode"
        }
    }
}

File: workspaces/ws_004_qrbarcodescanner/android/app/src/main/java/com/appfactory/qrbarcodescanner/scanner/BarcodeSymbology.kt
package com.appfactory.qrbarcodescanner.scanner

import com.google.mlkit.vision.barcode.common.Barcode

/**
 * Maps ML Kit barcode format constants to the symbology names declared in
 * the app spec, and provides the set of formats the scanner detects.
 */
object BarcodeSymbology {

    /**
     * Every barcode format the scanner listens for. Covers QR plus all 1D/2D
     * barcodes in the spec: EAN-13, EAN-8, UPC-A, UPC-E, Code 128, Code 39,
     * ITF, PDF417, Aztec, Data Matrix.
     */
    val supportedFormats = intArrayOf(
        Barcode.FORMAT_QR_CODE,
        Barcode.FORMAT_EAN_13,
        Barcode.FORMAT_EAN_8,
        Barcode.FORMAT_UPC_A,
        Barcode.FORMAT_UPC_E,
        Barcode.FORMAT_CODE_128,
        Barcode.FORMAT_CODE_39,
        Barcode.FORMAT_CODE_93,
        Barcode.FORMAT_ITF,
        Barcode.FORMAT_PDF417,
        Barcode.FORMAT_AZTEC,
        Barcode.FORMAT_DATA_MATRIX,
    )

    /** Human-readable name for an ML Kit barcode format constant. */
    fun displayName(format: Int): String = when (format) {
        Barcode.FORMAT_QR_CODE -> "QR Code"
        Barcode.FORMAT_EAN_13 -> "EAN-13"
        Barcode.FORMAT_EAN_8 -> "EAN-8"
        Barcode.FORMAT_UPC_A -> "UPC-A"
        Barcode.FORMAT_UPC_E -> "UPC-E"
        Barcode.FORMAT_CODE_128 -> "Code 128"
        Barcode.FORMAT_CODE_39 -> "Code 39"
        Barcode.FORMAT_CODE_93 -> "Code 93"
        Barcode.FORMAT_ITF -> "ITF"
        Barcode.FORMAT_PDF417 -> "PDF417"
        Barcode.FORMAT_AZTEC -> "Aztec"
        Barcode.FORMAT_DATA_MATRIX -> "Data Matrix"
        else -> "Barcode"
    }
}
